### ISSUE TITLE
Improve vtkMRMLDisplayNode API to support showing object in all view but one.

### ISSUE THREAD

**Issue by jcfr** [Slicer/Slicer#7434]
## Is your feature request related to a problem? Please describe.

The current API presents a challenge when attempting to selectively hide an object in only one view. This limitation necessitates the repetitive re-implementation of logic within modules and custom applications.

There are two cases to address:

1. **Hide in All Views Except One**: Users often need the flexibility to hide an object in all views except a specific one.

2. **Hide in Only the Specified View**: Similarly, there is a need to hide an object exclusively in a designated view. 

### Case (1): Hide in All Views Except One

To facilitate this, a new function, `vtkMRMLDisplayNode::HideOnlyInView(const char *viewNodeID)`[^1], could be introduced. The proposed implementation is as follows:

[^1]: name based of the existing API `vtkMRMLDisplayNode::SetDisplayableOnlyInView(const char *viewNodeID)`).

```cpp
  // Get all viewNodeIDs
  this->GetNumberOfNodesByClass("vtkMRMLAbstractViewNode");
  std::vector<vtkMRMLNode *> viewNodes;
  this->GetScene()->GetNodesByClass("vtkMRMLAbstractViewNode", viewNodes);
  std::vector<std::string> viewNodeIDs;
  for(vtkMRMLNode* viewNode: viewNodes)
    {
    viewNodeIDs.push_back(viewNode->GetID());
    }

  // Hide only in the specified view
  this->SetViewNodeIDs(viewNodeIDs);
  this->RemoveViewNodeID(viewNodeID);
```

### Case (2): Hide in Only the Specified View

Addressing this case is more complex due to the `ViewNodeIDs` convention. We lack a mechanism to distinguish between:

- **Intentional Hiding**: A view from which an object was explicitly hidden.
- **Subsequent Views**: For instance, switching to `Dual 3D` may imply the object should be visible by default.

Further refinement in the API or additional conventions may be necessary to address this.

**ViewNodeIDs Convention**:
- An empty list implies visibility in all views.
- A non-empty list indicates visibility exclusively in the specified view.

## Describe alternatives you've considered

NA

## Additional context

* https://github.com/Slicer/Slicer/pull/7433#discussion_r1406442044

**Comment by lassoan** [comment]
I agree, that the current single list of view node IDs cannot capture what the user wants (show only in the explicitly listed views; or hide in the non-listed views).

We could add a list of "blocked" view node IDs where an object cannot be displayed in. These would not be even offered in the GUI (and for some reason if a blocked view node ID would show up in the view node ID list, the node would not appear there).

Alternatively, we could add a node property that controls if the node should be displayed by default in new views. For this we may need to maintain a show and hide list for view node IDs (so that we can determine which are "new" views).

We could also specify view groups in addition to view node IDs (for example, "by default show the node in all view nodes with group ID = 1").

To choose between the difference approaches, it would be better to collect a list of use cases and their exact motivation for hiding an object in a view.

For example, if we need to hide a node in a certain view because other objects in that view are in a different coordinate system then this need would be much better addressed by specifying `FrameOfReferenceUID` for all views and displayable nodes; and only display those nodes in a view that have matching `FrameOfReferenceUID`. Maybe this would actually cover majority of the needs and we would not need any other mechanism for view node filtering.

**Comment by pieper** [comment]
Another option could be to keep the current list, but have a flag to invert the sense from "show in selected" to "hide in selected".

**Comment by cpinter** [comment]
> We could add a list of "blocked" view node IDs where an object cannot be displayed in. These would not be even offered in the GUI (and for some reason if a blocked view node ID would show up in the view node ID list, the node would not appear there).

I think this would work well. We already had similar lists in SH tree view filters (whitelists and blacklists, not sure what their new names are). It is easy to understand and do not add complexity really.

> Alternatively, we could add a node property that controls if the node should be displayed by default in new views. For this we may need to maintain a show and hide list for view node IDs (so that we can determine which are "new" views).

This is a good idea as well, regardless the decision about the block/negative list. This would make things a bit more complex though.

> To choose between the difference approaches, it would be better to collect a list of use cases and their exact motivation for hiding an object in a view.

My main use case is the panorama view in the dental application. There are some nodes that I want to show everywhere but in panorama, and there are other nodes (straightened clones) that I want to show only in panorama.
The Kitware folks are developing a dental application as well, so maybe their use case is the same (@jcfr please let me know if you'd like to join forces or have something not yet public to offer/sell in this domain).
